substrate/primitives/version/src/lib.rs

```diff
@@ -230,9 +221,394 @@ pub struct RuntimeVersion {
 	/// This number should never decrease.
 	pub transaction_version: u32,
 
-	/// Version of the state implementation used by this runtime.
+	/// Version of the system implementation used by this runtime.
 	/// Use of an incorrect version is consensus breaking.
-	pub state_version: u8,
+	pub system_version: u8,
+}
+
+// Manual implementation in order to sprinkle `stateVersion` at the end for migration purposes
+// after the field was renamed from `state_version` to `system_version`
+#[cfg(feature = "serde")]
+impl serde::Serialize for RuntimeVersion {
+	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+	where
+		S: serde::Serializer,
+	{
+		use serde::ser::SerializeStruct;
+
+		let mut state = serializer.serialize_struct("RuntimeVersion", 9)?;
+		state.serialize_field("specName", &self.spec_name)?;
+		state.serialize_field("implName", &self.impl_name)?;
+		state.serialize_field("authoringVersion", &self.authoring_version)?;
+		state.serialize_field("specVersion", &self.spec_version)?;
+		state.serialize_field("implVersion", &self.impl_version)?;
+		state.serialize_field("apis", {
+			struct SerializeWith<'a>(&'a ApisVec);
+
+			impl<'a> serde::Serialize for SerializeWith<'a> {
+				fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+				where
+					S: serde::Serializer,
+				{
+					apis_serialize::serialize(self.0, serializer)
+				}
+			}
+
+			&SerializeWith(&self.apis)
+		})?;
+		state.serialize_field("transactionVersion", &self.transaction_version)?;
+		state.serialize_field("systemVersion", &self.system_version)?;
+		state.serialize_field("stateVersion", &self.system_version)?;
+		state.end()
+	}
+}
+
+// Manual implementation in order to allow both old `stateVersion` and new `systemVersion` to be
+// present at the same time
+#[cfg(feature = "serde")]
+impl<'de> serde::Deserialize<'de> for RuntimeVersion {
+	fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+	where
+		D: serde::Deserializer<'de>,
+	{
+		use core::marker::PhantomData;
+
+		enum Field {
+			SpecName,
+			ImplName,
+			AuthoringVersion,
+			SpecVersion,
+			ImplVersion,
+			Apis,
+			TransactionVersion,
+			SystemVersion,
+			Ignore,
+		}
+
+		struct FieldVisitor;
+
+		impl<'de> serde::de::Visitor<'de> for FieldVisitor {
+			type Value = Field;
+
+			fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+				formatter.write_str("field identifier")
+			}
+
+			fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
```

Comment:
Do we need this?

Reply:
I didn't write it, I used macro expansion as the basis and this is what it spit out, I just made it a bit more readable